commit e0550222e03bae3fd629641e246ef7f47803d795
Author: Justin Stitt <justinstitt@google.com>
Date:   Mon Apr 29 23:06:54 2024 +0000

    printk: cleanup deprecated uses of strncpy/strcpy
    
    Cleanup some deprecated uses of strncpy() and strcpy() [1].
    
    There doesn't seem to be any bugs with the current code but the
    readability of this code could benefit from a quick makeover while
    removing some deprecated stuff as a benefit.
    
    The most interesting replacement made in this patch involves
    concatenating "ttyS" with a digit-led user-supplied string. Instead of
    doing two distinct string copies with carefully managed offsets and
    lengths, let's use the more robust and self-explanatory scnprintf().
    scnprintf will 1) respect the bounds of @buf, 2) null-terminate @buf, 3)
    do the concatenation. This allows us to drop the manual NUL-byte assignment.
    
    Also, since isdigit() is used about a dozen lines after the open-coded
    version we'll replace it for uniformity's sake.
    
    All the strcpy() --> strscpy() replacements are trivial as the source
    strings are literals and much smaller than the destination size. No
    behavioral change here.
    
    Use the new 2-argument version of strscpy() introduced in Commit
    e6584c3964f2f ("string: Allow 2-argument strscpy()"). However, to make
    this work fully (since the size must be known at compile time), also
    update the extern-qualified declaration to have the proper size
    information.
    
    Link: https://www.kernel.org/doc/html/latest/process/deprecated.html#strncpy-on-nul-terminated-strings [1]
    Link: https://github.com/KSPP/linux/issues/90 [2]
    Link: https://manpages.debian.org/testing/linux-manual-4.8/strscpy.9.en.html [3]
    Cc: linux-hardening@vger.kernel.org
    Signed-off-by: Justin Stitt <justinstitt@google.com>
    Reviewed-by: Petr Mladek <pmladek@suse.com>
    Link: https://lore.kernel.org/r/20240429-strncpy-kernel-printk-printk-c-v1-1-4da7926d7b69@google.com
    [pmladek@suse.com: Removed obsolete brackets and added empty lines.]
    Signed-off-by: Petr Mladek <pmladek@suse.com>

diff --git a/include/linux/printk.h b/include/linux/printk.h
index 2fde40cc9677..dbbd202b1cb3 100644
--- a/include/linux/printk.h
+++ b/include/linux/printk.h
@@ -71,7 +71,7 @@ extern void console_verbose(void);
 
 /* strlen("ratelimit") + 1 */
 #define DEVKMSG_STR_MAX_SIZE 10
-extern char devkmsg_log_str[];
+extern char devkmsg_log_str[DEVKMSG_STR_MAX_SIZE];
 struct ctl_table;
 
 extern int suppress_printk;